ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py

```diff
@@ -190,25 +149,129 @@ def remove(self, app, leaveObjects=False):
 
+    def update_object(self, app, parent, relname, object_id, spec, specparam):
+        """Compare object to be installed to existing objects, optionally creating the new object"""
+        # this backup should exist if previous installed version of this zenpack uses ZPL 2.0
+        backup_target = self.get_object(parent, relname, "{}-backup".format(object_id))
+        # otherwise this is the existing object pre-zpl 2.0
+        existing_target = self.get_object(parent, relname, object_id)
+        # we'll use whichever we get
+        candidate_target = backup_target or existing_target
+
+        if not candidate_target:
+            # if no candidate target is found, then this is new to this zenpack
+            spec.create(app.zport.dmd)
+        else:
+            # check the difference between our candidate and the new spec
+            # and back up the candidate if there is a difference
+            diff = self.check_diff(app, parent, relname, candidate_target, spec, specparam)
+            # if there was a difference, keep the backup and create the new template
+            if diff:
+                spec.create(app.zport.dmd)
+            else:
+                # otherwise just return the backup template to its original location
+                if backup_target:
+                    self.move_object(parent, relname, backup_target.id, object_id)
+                # or in the case of the existing object, leave it alone
+                else:
+                    pass
+
+    def check_diff(self, app, parent, relname, object, spec, specparam):
+        """Return True if object has changed creating preupgrade backup if needed"""
+        diff = self.object_changed(app, object, spec, specparam)
+        if diff:
+            # preserve the existing object if different
+            time_str = time.strftime("%Y%m%d%H%M", time.localtime())
+            preupgrade_id = "{}-preupgrade-{}".format(object.id, time_str)
+            self.move_object(parent, relname, object.id, preupgrade_id)
+            LOG.info("Existing object {}/{} differs from "
+                     "the newer version included with the {} ZenPack.  "
+                     "The existing object will be "
+                     "backed up to '{}'.  Please review and reconcile any "
+                     "local changes before deleting the backup:\n{}".format(
+                        parent.getDmdKey(), object.id, self.id, preupgrade_id, diff))
+            return True
+        return False
+
+    def get_object(self, parent, relname, object_id):
+        """Attempt to retrieve an object given its id, parent instance, and relation name"""
+        rel = getattr(parent, relname, None)
+        if rel:
+            try:
+                return rel._getOb(object_id)
+            except AttributeError:
+                pass
+        return None
+
+    def get_or_create_backup(self, parent, relname, object_id, overwrite=True):
```

Comment:
I don't see the overwrite argument actually being used. I'd stick to the [YAGN principle](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it) here and not create code paths that aren't used.